File: dotNet/astenn/astenn-core/src/Org/Lestr/Astenn/Configuration/XMLDocumentPersistenceDriver.cs
/*
 * Crée par SharpDevelop.
 * Utilisateur: Choupis
 * Date: 30/04/2010
 * Heure: 22:33
 * 
 * Pour changer ce modèle utiliser Outils | Options | Codage | Editer les en-têtes standards.
 */
using System;
using System.Collections.Generic;
using System.Xml;
using Org.Lestr.Astenn.Plugin;

namespace Org.Lestr.Astenn.Configuration
{
	
	
	public class XMLDocumentPersistenceDriver : IPersistenceDriver {
	
	
	    private XmlDocument document;
	
	
	    public XMLDocumentPersistenceDriver()
	    {
	
	    	document = new XmlDocument();
	    	
	    	document.AppendChild(document.CreateElement("astenn"));
	
	    }// END Constructor


	    public XMLDocumentPersistenceDriver(XmlDocument document)
	    {
	        
	        this.document = document;
	
	    }// END Constructor
	    
	    
		public XmlDocument Document 
		{
	    	
			get { return document; }
			set { document = value; }
			
		}// END Property Document
	
	
	    public void AddPluginInterface(string pluginInterfaceName) 
	    {
			
	    	XmlElement interfaceElement = Document.CreateElement("plugin-interface");
	    	interfaceElement.SetAttribute("class-name", pluginInterfaceName);
	    	
	    	document.DocumentElement.AppendChild(interfaceElement);
	
	    }// END Method AddPluginInterface
	
	
	    public void RemovePluginInterface(string pluginInterfaceName) 
	    {
	
	    	document.DocumentElement.RemoveChild(GetPluginInterfaceElement(pluginInterfaceName));
	
	    }// END Method RemovePluginInterface
	
	
	    public bool ExistPluginInterface(string pluginInterfaceName) 
	    {
	
	        return GetPluginInterfaceElement(pluginInterfaceName) != null;
	
	    }// END Method ExistPluginInterface


        public void AddPluginImplementation(string pluginInterfaceName, string pluginImplementationAddress) 
	    {
			
	    	XmlElement implementationElement = Document.CreateElement("plugin-implementation");
            implementationElement.SetAttribute("address", pluginImplementationAddress.ToString());
	    	
	    	GetPluginInterfaceElement(pluginInterfaceName).AppendChild(implementationElement);
	
	    }// END Method AddPluginImplementation


        public void RemovePluginImplementation(string pluginInterfaceName, string pluginImplementationAddress) 
	    {
	
	    	XmlElement pluginImplementationElement = GetPluginImplementationElement(pluginInterfaceName, pluginImplementationAddress);
	    	pluginImplementationElement.ParentNode.RemoveChild(pluginImplementationElement);
	
	    }// END Method RemovePluginImplementation


        public bool ExistPluginImplementation(string pluginInterfaceName, string pluginImplementationAddress) 
	    {
	
	    	return GetPluginImplementationElement(pluginInterfaceName, pluginImplementationAddress) != null;
	
	    }// END Method ExistPluginImplementation
	
	
	    public IEnumerable<string> GetPluginInterfacesNames() 
	    {
	
	    	List<string> rslt = new List<string>();
	    	
	    	foreach(XmlNode node in Document.DocumentElement.ChildNodes)
	    		if(node.NodeType == XmlNodeType.Element &&
	    		   node.Name.Equals("plugin-interface") &&
	    		   ((XmlElement)node).HasAttribute("class-name"))
	    			rslt.Add(((XmlElement)node).GetAttribute("class-name"));
	    	
	    	return rslt;
	
	    }// END Method GetPluginImplementationsNames


        public IEnumerable<string> GetPluginImplementationsAddresses(string pluginInterfaceName) 
	    {

            List<string> rslt = new List<string>();
	    	
	    	foreach(XmlNode node in GetPluginInterfaceElement(pluginInterfaceName).ChildNodes)
	    		if(node.NodeType == XmlNodeType.Element &&
	    		   node.Name.Equals("plugin-implementation") &&
                   ((XmlElement)node).HasAttribute("address"))
                    rslt.Add(((XmlElement)node).GetAttribute("address"));
	    	
	    	return rslt;
	
	    }// END Method GetPluginImplementationsURIs
	    
	    
	    private XmlElement GetPluginInterfaceElement(string pluginInterfaceName)
	    {
	    	
	    	foreach(XmlNode node in Document.DocumentElement.ChildNodes)
	    		if(node.NodeType == XmlNodeType.Element &&
	    		   node.Name.Equals("plugin-interface") &&
	    		   ((XmlElement)node).HasAttribute("class-name") &&
	    		   ((XmlElement)node).GetAttribute("class-name").Equals(pluginInterfaceName))
	    			return (XmlElement)node;
	    	
	    	return null;
	    	
	    }// END Method GetPluginInterfaceElement


        private XmlElement GetPluginImplementationElement(string pluginInterfaceName, string pluginImplementationURI)
	    {
	    	
	    	XmlElement pluginInterfaceElement = GetPluginInterfaceElement(pluginInterfaceName);
	    	
	    	if(pluginInterfaceElement != null)
		    	foreach(XmlNode node in pluginInterfaceElement.ChildNodes)
		    		if(node.NodeType == XmlNodeType.Element &&
		    		   node.Name.Equals("plugin-implementation") &&
                       ((XmlElement)node).HasAttribute("address") &&
                       ((XmlElement)node).GetAttribute("address").Equals(pluginImplementationURI.ToString()))
		    			return (XmlElement)node;
	    	
	    	return null;
	    	
	    }// END Method GetPluginInterfaceElement
	    

	}// END Classe XMLDocumentPersistenceDriver
	
	
}// END Namespace Astenn.Configuration

File: astenn-core/src/main/java/org/lestr/astenn/configuration/XMLDocumentPersistenceDriver.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package org.lestr.astenn.configuration;

import org.lestr.astenn.plugin.IPersistenceDriver;
import java.util.ArrayList;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

/**
 *
 * @author PIBONNIN
 */
public class XMLDocumentPersistenceDriver implements IPersistenceDriver {


    private Document document;


    public XMLDocumentPersistenceDriver() {

        try {
            document = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
            document.appendChild(document.createElement("astenn"));
        } catch (ParserConfigurationException ex) {
            Logger.getLogger(XMLDocumentPersistenceDriver.class.getName()).log(Level.SEVERE, null, ex);
        }

    }// END Constructor


    public XMLDocumentPersistenceDriver(Document document) {

        super();

        this.document = document;

    }// END Constructor


    /**
     * @return the document
     */
    public Document getDocument() {

        return document;

    }// END Method getDocument


    /**
     * @param document the document to set
     */
    public void setDocument(Document document) {

        this.document = document;

    }// END Method setDocument


    @Override
    public void addPluginInterface(String pluginInterfaceName) {

        Element interfaceElement = getDocument().createElement("plugin-interface");
        interfaceElement.setAttribute("class-name", pluginInterfaceName);

        document.getDocumentElement().appendChild(interfaceElement);

    }// END Method addPluginInterface


    @Override
    public void removePluginInterface(String pluginInterfaceName) {

        document.getDocumentElement().removeChild(getPluginInterfaceElement(pluginInterfaceName));

    }// END Method removePluginInterface


    @Override
    public boolean existPluginInterface(String pluginInterfaceName) {

        return getPluginInterfaceElement(pluginInterfaceName) != null;

    }// END Method existPluginInterface


    @Override
    public void addPluginImplementation(String pluginInterfaceName,
                                        String pluginImplementationAddress) {

        Element implementationElement = getDocument().createElement("plugin-implementation");
        implementationElement.setAttribute("address", pluginImplementationAddress);

        getPluginInterfaceElement(pluginInterfaceName).appendChild(implementationElement);

    }// END Method addPluginImplementation


    @Override
    public void removePluginImplementation(String pluginInterfaceName,
                                           String pluginImplementationAddress) {

        Element pluginImplementationElement = getPluginImplementationElement(pluginInterfaceName, pluginImplementationAddress);
        pluginImplementationElement.getParentNode().removeChild(pluginImplementationElement);

    }// END Method removePluginImplementation


    @Override
    public boolean existPluginImplementation(String pluginInterfaceName,
                                             String pluginImplementationAddress) {

        return existPluginInterface(pluginInterfaceName) && getPluginImplementationElement(pluginInterfaceName, pluginImplementationAddress) != null;

    }// END Method existPluginImplementation


    @Override
    public Iterable<String> getPluginInterfacesNames() {

        ArrayList<String> rslt = new ArrayList<String>();

        NodeList children = getDocument().getDocumentElement().getChildNodes();

        for (int i = 0; i <= children.getLength() - 1; i = i + 1) {

            Node node = children.item(i);

            if (node.getNodeType() == Node.ELEMENT_NODE
                && node.getNodeName().equals("plugin-interface")
                && ((Element) node).hasAttribute("class-name"))
                rslt.add(((Element) node).getAttribute("class-name"));

        }

        return rslt;

    }// END Method getPluginImplementationsNames


    @Override
    public Iterable<String> getPluginImplementationsAddresses(String pluginInterfaceName) {

        ArrayList<String> rslt = new ArrayList<String>();

        Element pluginInterfaceElement = getPluginInterfaceElement(pluginInterfaceName);

        if (pluginInterfaceElement != null) {

            NodeList children = pluginInterfaceElement.getChildNodes();

            for (int i = 0; i <= children.getLength() - 1; i = i + 1) {

                Node node = children.item(i);

                if (node.getNodeType() == Node.ELEMENT_NODE
                    && node.getNodeName().equals("plugin-implementation")
                    && ((Element) node).hasAttribute("address"))
                    rslt.add(((Element) node).getAttribute("address"));

            }

        }

        return rslt;

    }// END Method getPluginImplementationsStrings


    private Element getPluginInterfaceElement(String pluginInterfaceName) {

        NodeList children = getDocument().getDocumentElement().getChildNodes();

        for (int i = 0; i <= children.getLength() - 1; i = i + 1) {

            Node node = children.item(i);

            if (node.getNodeType() == Node.ELEMENT_NODE
                && node.getNodeName().equals("plugin-interface")
                && ((Element) node).hasAttribute("class-name")
                && ((Element) node).getAttribute("class-name").equals(pluginInterfaceName))
                return (Element) node;

        }

        return null;

    }// END Method getPluginInterfaceElement


    private Element getPluginImplementationElement(String pluginInterfaceName,
                                                   String pluginImplementationAddress) {

        Element pluginInterfaceElement = getPluginInterfaceElement(pluginInterfaceName);

        if (pluginInterfaceElement != null) {

            NodeList children = getDocument().getDocumentElement().getChildNodes();

            for (int i = 0; i <= children.getLength() - 1; i = i + 1) {

                Node node = children.item(i);

                if (node.getNodeType() == Node.ELEMENT_NODE
                    && node.getNodeName().equals("plugin-implementation")
                    && ((Element) node).hasAttribute("address")
                    && ((Element) node).getAttribute("address").equals(pluginImplementationAddress))
                    return (Element) node;

            }

        }

        return null;

    }// END Method getPluginInterfaceElement


}// END Class XMLDocumentPersistenceDriver


